### ISSUE TITLE
How to train to detect plate number?

### ISSUE THREAD

**Issue by masoudnick** [AlexeyAB/darknet#2468]
hi,I want to train a plate number detection but I do not know what model and cfg file to use?
How much batch,subdivisions,width and height should be?
How much number of images should be?
What should I do to get the best result?
Please help me :)

**Comment by Jacob-Stevens-Haas** [comment]
There's always a bit of guessing with neural nets, but here's my recommendation:

batch: keep at 64
subdivisions: depends on memory of your GPU.  Keep as low as possible (1), but if you run out of memory, increase to 2,4, 8, 16, 32, or 64.
width & height: close to whatever your image data is, but needs to be a multiple of 32.  There's been some issues posted with non-square dimensions, but I've gotten good results.  After you train with one set of width and height, try using square dimensions and see if your detection improves
number of images: There's no hard and fast rule.  I'd say start with 500-1000 and see if that's good enough, if you can get that many.  If collecting images and tagging them is taking too long, try with less. 
 If your data also has things that look like license plates (e.g. road signs that are the same size and shape) you might need more training images.

**Comment by masoudnick** [comment]
@Jacob-Stevens-Haas 
Thanks for your response :)
I collected about 500 images of license plates and I did all the same steps as for the tiny-yolo model as described in github
But I always come up with -nan avg loss message during the training (usually.training until fifty iterations goes well).
Can you guide me please? What should I do?

content of yolov3-tiny.cfg:
batch=64
subdivisions=2
width=416
height=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

**Comment by Jacob-Stevens-Haas** [comment]
All of that looks good.  My guess is that it has to do with how your training data is set up.  Did you use YoloMark?

Also, could you paste the batch before and after you get a -nan avg loss?  Darknet pauses if you click in the window.  For example, this is most of one batch.

``` 
10777: 2.423840, 2.634885 avg loss, 0.000100 rate, 11.249701 seconds, 689728 images
Loaded: 0.000000 seconds
Region 16 Avg IOU: -nan(ind), Class: -nan(ind), Obj: -nan(ind), No Obj: 0.000011, .5R: -nan(ind), .75R: -nan(ind),  count: 0
Region 23 Avg IOU: 0.644044, Class: 0.995289, Obj: 0.536937, No Obj: 0.001538, .5R: 1.000000, .75R: 0.000000,  count: 4
Region 16 Avg IOU: -nan(ind), Class: -nan(ind), Obj: -nan(ind), No Obj: 0.000014, .5R: -nan(ind), .75R: -nan(ind),  count: 0
Region 23 Avg IOU: 0.659605, Class: 0.962077, Obj: 0.600405, No Obj: 0.004886, .5R: 0.833333, .75R: 0.333333,  count: 12
Region 16 Avg IOU: 0.899500, Class: 0.999506, Obj: 0.448131, No Obj: 0.001220, .5R: 1.000000, .75R: 1.000000,  count: 1
Region 23 Avg IOU: 0.813568, Class: 0.999372, Obj: 0.804746, No Obj: 0.001612, .5R: 1.000000, .75R: 0.666667,  count: 3
Region 16 Avg IOU: -nan(ind), Class: -nan(ind), Obj: -nan(ind), No Obj: 0.000070, .5R: -nan(ind), .75R: -nan(ind),  count: 0
Region 23 Avg IOU: 0.587234, Class: 0.845241, Obj: 0.426603, No Obj: 0.004455, .5R: 0.714286, .75R: 0.285714,  count: 14
Region 16 Avg IOU: -nan(ind), Class: -nan(ind), Obj: -nan(ind), No Obj: 0.000036, .5R: -nan(ind), .75R: -nan(ind),  count: 0```

**Comment by masoudnick** [comment]
@Jacob-Stevens-Haas  
Yes, i used YoloMark and labeled each object on images with great care.

batch before -nan avg:

31: 1238.394043, 926.503906 avg loss, 0.000000 rate, 435.952099 seconds, 1984 images
Loaded: 0.000000 seconds
Region 16 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 7
Region 23 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 6
Region 16 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 8
Region 23 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 2
Region 16 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 6
Region 23 Avg IOU: -nan(ind), Class: -nan(ind), Obj: -nan(ind), No Obj: -nan, .5R: -nan(ind), .75R: -nan(ind),  count: 0
Region 16 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 8
Region 23 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 3
Region 16 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 7
Region 23 Avg IOU: -nan(ind), Class: -nan(ind), Obj: -nan(ind), No Obj: -nan, .5R: -nan(ind), .75R: -nan(ind),  count: 0
Region 16 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 8
Region 23 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 16 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 5
Region 23 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 5
Region 16 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 6
Region 23 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 13


batch after-nan avg:

32: -nan, -nan avg loss, 0.000000 rate, 389.016436 seconds, 2048 images
Loaded: 0.000000 seconds
Region 16 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 7
Region 23 Avg IOU: nan, Class: nan, Obj: nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 2
Region 16 Avg IOU: -nan, Class: nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 5
Region 23 Avg IOU: nan, Class: -nan, Obj: nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 3
Region 16 Avg IOU: nan, Class: -nan, Obj: nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 8
Region 23 Avg IOU: -nan(ind), Class: -nan(ind), Obj: -nan(ind), No Obj: -nan, .5R: -nan(ind), .75R: -nan(ind),  count: 0
Region 16 Avg IOU: nan, Class: -nan, Obj: nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 6
Region 23 Avg IOU: nan, Class: nan, Obj: nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 3
Region 16 Avg IOU: nan, Class: nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 7
Region 23 Avg IOU: nan, Class: nan, Obj: nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 2
Region 16 Avg IOU: nan, Class: nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 7
Region 23 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 16 Avg IOU: -nan, Class: nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 7
Region 23 Avg IOU: -nan, Class: -nan, Obj: nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 16 Avg IOU: nan, Class: nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 9
Region 23 Avg IOU: nan, Class: -nan, Obj: -nan, No Obj: -nan, .5R: 0.000000, .75R: 0.000000,  count: 1

**Comment by Favi0** [comment]
im working on a similar project 

First of all : did you configure everything data.obj, names.obj and cfg ?
what is your img resolution ?
Which architecture are you using? are you training from scratch or not? based on my experience i would tell you to get a set of pretrained weights from coco (like the tutorial) and train from there , you have only 500 images , thats OK for transfer learning for 1 class (licence class)

**Comment by masoudnick** [comment]
@Favi0 
Thanks for your response :)
Yes,i configured all of these files.
My images resolution is very different.
I am using from yolo-tiny and training from scratch but I have not yet reached the conclusion ):
Which which pretrained weights do you suggest?
Explain more.thank you.

**Comment by AlexeyAB** [comment]
@masoudnick 

Use this pre-trained wights file ` yolov3-tiny.conv.15` that you can get by using these commands:
https://github.com/AlexeyAB/darknet#how-to-train-tiny-yolo-to-detect-your-custom-objects

**Comment by masoudnick** [comment]
@AlexeyAB 
Ok,but i dont know why always come up with -nan avg before fifty iterations.

**Comment by AlexeyAB** [comment]
@masoudnick 

Do you still get `nan` for Avg loss even with `yolov3-tiny.conv.15` ?

Do you have files bad.list or bad_label.list?

**Comment by masoudnick** [comment]
@AlexeyAB 
Yes,as I said,i always got `nan` for Avg loss before fifty iterations (with `yolov3-tiny.conv.15`) and dont have any file like bad.list.

**Comment by masoudnick** [comment]
@AlexeyAB 
I did run `calc_mAP.cmd` and i got this results:

```
 detections_count = 1477, unique_truth_count = 628
class_id = 0, name = plate,      ap = 0.00 %
 for thresh = 0.25, precision = 0.00, recall = 0.00, F1-score = -nan(ind)
 for thresh = 0.25, TP = 0, FP = 2, FN = 628, average IoU = 0.00 %
 IoU threshold = 50 %
 mean average precision (mAP@0.50) = 0.000000, or 0.00 %
Total Detection Time: 273.000000 Seconds
```

can you explain about this results please?

**Comment by AlexeyAB** [comment]
@masoudnick Hi,

Your model is trained incorrectly.

* Rename your cfg-file to txt-file and drag-n-drop to your message.
* Do you use the latest version of Darknet?
* What parameters in the Makefile do you use?
* What command do you use for training?

**Comment by masoudnick** [comment]
@AlexeyAB  Hi,

cfg-file:
[YOLO3-tiny-obj.txt](https://github.com/AlexeyAB/darknet/files/2909753/YOLO3-tiny-obj.txt)

I installed darknet by this command:
`git clone https://github.com/pjreddie/darknet`

I dont know anything about Makefile.
training command:
`darknet_no_gpu.exe detector train tiny-obj.data yolo3-tiny-obj.cfg yolov3-tiny.conv.15 -dont_show`

**Comment by AlexeyAB** [comment]
@masoudnick 

> cfg-file:
> [YOLO3-tiny-obj.txt](https://github.com/AlexeyAB/darknet/files/2909753/YOLO3-tiny-obj.txt)

Train with default anchors instead of your anchors.

Yolov3-tiny should use only 6  anchors (12 values) instead of 9 anchors (18 values).

----

> I installed darknet by this command:
> `git clone https://github.com/pjreddie/darknet`

This is original repository that doesn't support Windows.

> training command:
> `darknet_no_gpu.exe detector train tiny-obj.data yolo3-tiny-obj.cfg yolov3-tiny.conv.15 -dont_show`

This is the command of Windows version of Darknet, that can't work with original repository.

**Comment by masoudnick** [comment]
> 
> 
> @masoudnick
> 
> > cfg-file:
> > [YOLO3-tiny-obj.txt](https://github.com/AlexeyAB/darknet/files/2909753/YOLO3-tiny-obj.txt)
> 
> Train with default anchors instead of your anchors.
> 
> Yolov3-tiny should use only 6 anchors (12 values) instead of 9 anchors (18 values).
> 
> > I installed darknet by this command:
> > `git clone https://github.com/pjreddie/darknet`
> 
> This is original repository that doesn't support Windows.
> 
> > training command:
> > `darknet_no_gpu.exe detector train tiny-obj.data yolo3-tiny-obj.cfg yolov3-tiny.conv.15 -dont_show`
> 
> This is the command of Windows version of Darknet, that can't work with original repository.

@AlexeyAB , Hi
as you said, I trained with default anchors, but still i have same problem. what should i do?

**Comment by AlexeyAB** [comment]
@masoudnick 

Do you have nVidia GPU and CUDA?

Try to use this command:
`darknet.exe detector train tiny-obj.data yolo3-tiny-obj.cfg yolov3-tiny.conv.15 -dont_show`

**Comment by KO-KMS-AT12** [comment]
> @Jacob-Stevens-Haas
> Thanks for your response :)
> I collected about 500 images of license plates and I did all the same steps as for the tiny-yolo model as described in github
> But I always come up with -nan avg loss message during the training (usually.training until fifty iterations goes well).
> Can you guide me please? What should I do?
> 
> content of yolov3-tiny.cfg:
> batch=64
> subdivisions=2
> width=416
> height=416
> channels=3
> momentum=0.9
> decay=0.0005
> angle=0
> saturation = 1.5
> exposure = 1.5
> hue=.1

Why channels = 3 ? When I trained to recognize letters from license plate number to channel 3, I could not recognize and channel 1 was recognized. Can you explain for me please?
